### DISCUSSION TITLE
Opinion on using Paste Shematic for pasting in ~13 million blocks?

### DISCUSSION THREAD

**Discussion by toydotgame** [maruohon/litematica#828] [General]
I have a thoroughly messed up rectangle of 1.6 underground terrain where ore distribution is really messed up and there's no cave air blocks. Case in point: 
![image](https://github.com/maruohon/litematica/assets/64190512/a67b5994-b2ba-4e1c-9e06-9795957ed901)

WE is too fragile to work with copying, pasting, and schematic-ising the roughly 550x420x60 (13.86 million m<sup>3</sup>) block area; additionally, I wanna use Litematica because there are parts I need to add complex shapes (likely through schematic sub-areas) where I'm not touching stuff because builds are already there.

How good is Litematica at handling pasting such large segments? I'd likely do this in singleplayer so server-related issues don't need to be considered.

**Comment by maruohon** [comment]
In single player it should be totally fine. There isn't any real upper limit to pasting, it's mostly "limited" by  the usual "how much memory does the schematic take to load in the first place" and "how laggy is the rendering" type of limitations. For better rendering performance with large schematics you can disable the colored schematic overlay (Visuals -> `enableSchematicOverlay`), or try the `overlayReducedInnerSides` optimization option, although that option is still buggy and unfinished. And also disable Generic -> `renderThreadNoTimeout`.

The pasting always happens in per-chunk pieces in the "recent" mod versions, so the schematic size is not really relevant to pasting. And the worst part of pasting itself is light updates, but pasting underground those are quite minimal. Light update issues are mostly relevant when pasting large things high up in an empty-ish world so that lots of light updates need to propagate down.

Just make sure to keep a backup of the world, as Litematica still doesn't have any undo functionality if you accidentally paste something wrong/in the wrong place etc.

Also for selective pasting in some cases you may want to use the `pasteReplaceBehavior` option (cycle it with `Ctrl + M` while in the Paste mode), depending on what and how you are pasting and if you need to kinda insert something inside the ground without carving out air cubes or replacing existing things.

**Reply by toydotgame** [reply]
Thanks! I'll definitely look into seeing how to best utilise `pasteReplaceBehavior` to do this job. Other than that, I think everything's pretty set now.

Somewhat related: how does Litematica handle biomes when pasting? I know—unfortunately from experience—that MCEdit replaces the entire Y column of blocks with the source copy's biome information (even if the copy region is like 5 blocks tall), and that WorldEdit has options to copy or to not copy biome information – but I'm not sure how Litematica handles it/if at all it does.

**Reply by maruohon** [reply]
Litematica doesn't do anything at all with biomes. (Meaning that it also doesn't save any biomes when creating schematics.)

It's *possible* that it might support biomes at some point in the future, but I'd say it's more likely that it won't. Because biomes are almost never relevant in its typical use cases (in a way that the mod should store and restore them - it would only be relevant in the Paste mode, nowhere else, I think).

**Reply by toydotgame** [reply]
Did the transfer today and it went very smoothly (I ended up not even disabling overlay rendering because after the initial loading lagspike I was still getting my usual average fps!)

(Conveniently also WorldEdit must've updated recently because I was able to `//gmask` and `//setbiome` hundreds of millions of blocks with no issue despite me trying 6 months ago and it dying at only a few hundred thousand)
